BUG: Make sure that _binary_erosion only accepts an integer number of iterations.

nateyoder opened the pull request:
Addresses #9905 and #9909 

Avoids the resulting memory error by raising on non-integer inputs by using operator.index

rfezzani commented:
I would recommend  to test if the argument `iterations` is a whole integer and raise a `ValueError` if it is not the case.
In fact, when iterations is a float `operation.index(iterations)` raises a `TypeError: 'float' object cannot be interpreted as an index`.
Testing if an argument is a whole integer can be made using the `float.is_integer` operator.

nateyoder commented:
Sounds good.  I'll change it over.  I had used the`operator.index(iterations)` based off this comment: https://github.com/scipy/scipy/pull/9392#discussion_r226612519 by @larsoner

rgommers commented:
> Testing if an argument is a whole integer can be made using the `float.is_integer` operator.

That's not quite right, doesn't work unless something is actually a float. `float.is_integer(2)` gives `TypeError`.

> In fact, when iterations is a float operation.index(iterations) raises a TypeError: 'float' object cannot be interpreted as an index.

This is the expected behavior I think. Users should be providing integers, not floats that happen to be integer-valued.

rfezzani commented:
A correct test would be:

```python
if float(iterations).is_integer():
  iterations = int(iterations)
else:
   raise ValueError('iterations should be a whole integer.')
```

rgommers commented:
I think `operator.index(iterations)` is the better thing to use here - it's a one-liner, and gives `TypeError` which seems right. We don't really want to allow float input.

rfezzani commented:
I have no real objection for the use of `operator.index(iterations)`, apart the fact that the raised `TypeError` comes from `operator.index` and not from `_binary_erosion`.
In this case, may be a more explicit error management

```python
try:
  iterations = operator.index(iterations)
except TypeError:
  raise TypeError(iteration parameter should be an integer)
```

It is no more a one-liner, but the raised error is much more localized and explicit.

rgommers commented:
fair enough, that is nicer

nateyoder commented:
Thanks for the feedback.  I updated as suggested above in https://github.com/scipy/scipy/pull/10009#issuecomment-481156587

rgommers commented:
Had a closer look, and this doesn't seem to address all of https://github.com/scipy/scipy/issues/9905#issuecomment-470938221. 

@nateyoder would you be able to also look at the `brute_force` argument and/or the C code?

nateyoder commented:
@rgommers thanks for catching that.  I updated the C code as suggested in the comment, removed the and that fix seemed to handle the memory error so I also did that in the other places in ndimage that had the same pattern.  I also added a test for brute_force and it didn't seem to be an issue as it isn't actually passed into the C function.

rgommers commented:
Merged, thanks @nateyoder. And thanks @rfezzani for reviewing.

nateyoder commented:
Thanks to you all for your patience, help and suggestions!  It was fun to contribute.

rgommers commented:
> Thanks to you all for your patience, help and suggestions! It was fun to contribute.

Glad to hear that. We'll be happy to see you contribute again:)
